Melhoria da categoria exceptions (#37)
paulorievrs/java4noobs: 3 - Básico/17-Excecoes.md

<h1>Exceções</h1>

> Geralmente ao construir um programa somos deparados com alguns erros no fluxo das nossas aplicações, mas não se preocupe com esse tópico você irá aprender um pouco mais sobre elas, e com isso chegamos a este tópico sobre exceções, aqui falaremos um pouco mais sobre, o que é, e como tornar possível com que sejam ignorados ou tratado esses erros caso seja o propósito do seu programa.

Author: Texto reformulado, tentei transformar em um texto mais introdutório ao módulo.

<h2>Tipos de Exceções</h2>

Primeiro vamos conhecer os tipos de exceções que temos no Java, iremos falar sobre alguns conceitos de POO que estão ligados aos tipos de exceções, mas não se preocupe que vamos conhecer mais sobre POO no próximo modulo intermediário.

<img style="margin: 20px 0 20px 0;" src="../img/arvore_exceptions.jpeg">

<h3>Throwable</h3>

Considerada a classe mãe das exceções, ou seja, toda exceção ou erro é considerada uma subclasse de Throwable.

<br>

<h3>Error</h3>

Erros sempre vai acontecer em tempo de execução e é uma condição considerada irrecuperável, ou seja, a sua aplicação é encerrada após tomar um throwable do tipo erro. Não pode ser tratada com blocos Try-catch pois os erros estão ligados a JVM normalmente são considerado "problemas" de baixo nível, como falta de memória, falhas de hardware ou problemas no sistema operacional.

<h3>Alguns exemplos de Throwable do tipo "Error"</h3>

> > OutOfMemoryError -> Esse erro se dá quando a JVM não consegue mais alocar nada na memória.

Reviewer: ```suggestion
> OutOfMemoryError -> Esse erro se dá quando a JVM não consegue alocar mais memória.
```

> > StackOverflowError ->  Acontece quando a quantidade de memória da pilha é ultrapassada.

Reviewer: ```suggestion
> StackOverflowError -> Acontece quando a quantidade de memória da pilha é ultrapassada.
```

> > NoClassDefFoundError -> Diz que a JVM tentou buscar internamente a pela sua classe e não a encontrou.

Reviewer: ```suggestion
> NoClassDefFoundError -> Ocorre quando a JVM não consegue encontrar a sua classe.
```

<br>

<h3>Exceptions</h3>

> As exceções são problemas que podem ocorrer durante a execução de um programa Java, mas não causam um colapso total do programa, e podem ser previstas e tratadas através dos blocos de exceção. As exceções são divididas em duas categorias: exceções verificadas (checked exceptions) e exceções não verificadas (unchecked exceptions). As exceções verificadas são aquelas que exigem tratamento explícito, ou seja, necessitam da utilização dos blocos de exceções, enquanto as exceções não verificadas não requerem tratamento explícito ou a utilização dos blocos de exceções, mas é considerado boa pratica a utilização pois casa não tenha o bloco Try-Catch o seu problema se encerrara com a exception.

Reviewer: ```suggestion
As exceções são problemas que podem ocorrer durante a execução de um programa Java, mas não causam um colapso total do programa, e podem ser previstas e tratadas através dos blocos de exceção. As exceções são divididas em duas categorias: exceções verificadas (checked exceptions) e exceções não verificadas (unchecked exceptions). As exceções verificadas são aquelas que exigem tratamento explícito, ou seja, necessitam da utilização dos blocos de exceções, enquanto as exceções não verificadas não requerem tratamento explícito ou a utilização dos blocos de exceções, mas é considerado boa pratica a utilização, pois caso não tenha o bloco Try-Catch, o seu programa se encerrará com a exception.
```

Author: Sugestões aceitas e adicionei que as exceções não verificadas são as RuntimeException.

<h3>Alguns exemplos de Exceptions do tipo "Checked"</h3>

> FileNotFoundException -> Ocorre quando um arquivo não pode ser encontrado durante uma leitura.

> SQLException -> Ocorre quando estamos lidando a nível de banco de dados, pode ser de erro de conexão, consultas invalidas pelo banco etc.

<h3>Alguns exemplos de Exceptions do tipo "Unchecked"</h3>

> ArrayIndexOutOfBoundsException -> Quando tenta acessar um index do array no qual não existe.

> NullPointerException -> Talvez a exception mais comum de acontecer, quando tentamos acessar ou executar alguma referencia (objeto) que esteja null ou seja nunca foi inicializada.

<h2>Criando suas exceções</h2>

> Você também pode criar suas exceções no Java para algum determinado fluxo ou alguma regra de negócio que você tenha em seu código, assim facilitando a localização e descobrindo o que pode ter acontecido pelo caminho, basta você criar uma classe estendendo a classe de erro.

Reviewer: ```suggestion
Você também pode criar suas exceções no Java para algum determinado fluxo ou alguma regra de negócio que possa disparar algum erro, assim facilitando a localização e descobrindo o que pode ter acontecido pelo caminho. Para isso, basta criar uma classe que herde alguma classe de erro, como `RuntimeException` ou `Exception`.
```
